### ISSUE TITLE
Strange memory leak detected with valgrind

### ISSUE THREAD

**Issue by danfloyd111** [Microsoft/WSL#2688]
* Windows build number:  10.0.16299.64
* lsb_release:
`
No LSB modules are available.
Distributor ID: Ubuntu
Description:     Ubuntu 16.04.03 LTS
Release:           16.04
Codename:      xenial
`

Compiling this simple program:
`
// file foo.cpp
#include <iostream>
 int main() {
   return EXIT_SUCCESS;
 }
`
with `g++ -std=c++17 foo.cpp -o foo.run`
and tested with `valgrind ./foo.run`

will produce this report on my machine (on Bash in Ubuntu on Windows):
`[...] in use at exit: 72,704 bytes in 1 blocks     total heap usage: 1 allocs, 0 frees 72,704 bytes allocated [...]`

If I compile the same code directly with g++ on Ubuntu, valgrind says that there aren't any leaks as expected (1 allocs, 1 frees).

Could anyone explain to me why this happens? It's my fault?
Thanks folks :)

**Comment by therealkenc** [comment]
Interesting one. Need `strace -ff valgrind ./foo.run 2> foo.strace` for both. Link gists if they are long. Or just look yourself to see if any differences stand out.

If you skip the `#include` does it still leak? Same g++ and libg++ on WSL and Real Ubuntu? Or, is the binary byte-for-byte the same, and if not, if you copy it over, does it still leak?

**Comment by danfloyd111** [comment]
*I've tested this also with SuseLeap on WSL wich runs g++ 4.8.5 and everything works as expected.

*On Ubuntu on WSL I have g++ 5.4.0
If I skip `#include <iostream>` it not compiles with `g++ std=c++17 foo.cpp -o foo.run`
If I substitute `return EXIT_SUCCESS` with `return 0` it compiles and runs with no leaks.

*On "real" Ubuntu I have g++ 7.2.0

Maybe the problem is in that particular version of g++ ? I'll investigate asap on the straces.

**Comment by therealkenc** [comment]
> I've tested this also with SuseLeap on WSL wich runs g++ 4.8.5 and everything works as expected.

If by "as expected" you mean "no leak", that points strongly to a library/toolchain difference rather than a kernel problem. But to be completely sure....

> If I substitute return EXIT_SUCCESS with return 0 it compiles and runs with no leaks.

The idea here is that including `<iostream>` allocates stuff on the heap. And indeed it looks like it does, if _not_ including `<iostream>` "cures" the leak.

But before you get too embroiled in `strace` you need a common point of comparison. This is because yeah it _could_ be a particular version of `g++`  or `<iostream>` thing. But even if it is not, you need to be able to compare the `strace` output same-for-same to learn where the difference (if any) lies.

Copy over the 5.4.0 generated binary from WSL to Real Ubuntu and see if it leaks there. If it leaks on one but not the other, that's when to start pouring over strace logs.

**Comment by danfloyd111** [comment]
> Copy over the 5.4.0 generated binary from WSL to Real Ubuntu and see if it leaks there. If it leaks on one but not the other, that's when to start pouring over strace logs.

Done, no leaks detected in real ubuntu nor in SuseLeap on WSL.
Should I compare the straces as next move?

Thank you so much for the help!

**Comment by therealkenc** [comment]
Okay so the 5.4.0 binary _only_ leaks on WSL Ubuntu 16.04? [But nothing else you've tried.] Do I have that right?

You can post/link the straces, but I think that might be excessive effort. From your description, after pouring over the strace, we're going to discover "yep it leaks right there". If you pick apart your Real Ubuntu and SuseLeap with a fine tooth comb, there's gotta be a library difference them and your WSL Ubuntu userspace.

You could _try_ compiling your test case with `g++ -static` on Real Ubuntu, copy the binary to WSL, and see if that leaks. But I haven't tried a statically linked cpp binary in 15 years.

In short you can chase this one further, but you probably aren't chasing a WSL problem. If it matters other than the curiosity factor, that leak is almost certainly going to be a one-off global pointer that has no practical effect, because it gets allocated at startup and will be reaped on process `_exit()`.

**Comment by danfloyd111** [comment]
> Okay so the 5.4.0 binary only leaks on WSL Ubuntu 16.04? [But nothing else you've tried.] Do I have that right?
Correct.

> In short you can chase this one further, but you probably aren't chasing a WSL problem. 
I agree with you.

Thank you very much for the support, I was very curious about this "strange" thing, first I was chasing the leak thinking that was hiding in my code and when I teared apart all the code and came up with just a few standard code lines I wasn't able to isolate more the above-mentioned behaviour. I think I'll go for g++ 7 on the WSL too and leave the straces apart.
